frontend/apps/reader/modules/readertypography.lua

```diff
@@ -770,14 +770,14 @@ function ReaderTypography:onReadSettings(config)
     self.allow_doc_lang_tag_override = false
     -- Use the one manually set for this document
     self.text_lang_tag = config:readSetting("text_lang")
-    if self.text_lang_tag then
+    if config:has("text_lang") then
         logger.dbg("Typography lang: using", self.text_lang_tag, "from doc settings")
         self.ui.document:setTextMainLang(self.text_lang_tag)
         return
     end
     -- Use the one manually set as default (with Hold)
     self.text_lang_tag = G_reader_settings:readSetting("text_lang_default")
-    if self.text_lang_tag then
+    if G_reader_settings:has("text_lang_default") then
```

Comment:
Worth it, looking twice at G_reader_settings ?
I find it a bit more obscure than the previous code.
Same in a next commit with a docsetting config:
```diff
    self.show_page_labels = config:readSetting("pagemap_show_page_labels")
-   if self.show_page_labels == nil then
+   if config:hasNot("pagemap_show_page_labels") then
        self.show_page_labels = G_reader_settings:nilOrTrue("pagemap_show_page_labels")
    end
```
Or remove the line before and do the read in the `then`. But even then... It's not needed if you use :nilOrTrue, or redundant.
And in the text_lang case, I'd rather explicitely assign something, even if nil, than just not assign anything (and not override stuff that would previously be there, even if it might not happen in this context).

Reply:
I find that it actually more clearly indicates the intent:

e.g., DocSetting doesn't contain that key => check elsewhere.

Reply:
I'm not sure I get the text_lang comment, because it doesn't affect assignments at all :?.

(And nearly all of those should be functionally identical cosmetic tweaks).

Reply:
Another way to write (most of) those DocSettings -> G_reader_settings constructs could perhaps be made clearer with:

```
if doc:has(key) then
  self.var = doc:readSetting(key)
else
  self.var = G:readSetting(key)
end
```

Reply:
> I find that it actually more clearly indicates the intent:
e.g., The DocSettings object doesn't contain that key => check elsewhere.

> Speaking of, (most of) those DocSettings -> G_reader_settings constructs could perhaps be made clearer with something like:

Yes, that's clearer, in that "DocSettings -> G_reader_settings"  case.

But the ones like this one are not clearer:
```
1    self.text_lang_tag = G_reader_settings:readSetting("text_lang_fallback")
2    if G_reader_settings:has("text_lang_fallback") then
3        self.ui.document:setTextMainLang(self.text_lang_tag)
4        return
5    end
```

1 Read a value from the far away setting
2 Check again if there is a value in the far away setting
3 Use the value fetched at 1

I find it more explicite as it was:
```
1    self.text_lang_tag = G_reader_settings:readSetting("text_lang_fallback")
2    if self.text_lang_tag then
3        self.ui.document:setTextMainLang(self.text_lang_tag)
4        return
5    end
```

1 Read a value from the far away setting
2 Check that the value is usable
3 Use this value

> I'm not sure I get the text_lang comment, because it doesn't affect assignments at all :?

I meant that _if_ you were to remove/move the step 1 in this sample to not have 2 calls (like you have done elsewhere), you'd get:

```
2    if G_reader_settings:has("text_lang_fallback") then
1        self.text_lang_tag = G_reader_settings:readSetting("text_lang_fallback")
3        self.ui.document:setTextMainLang(self.text_lang_tag)
4        return
5    end
```
You would not be reseting an existing `self.text_lang_tag = "tagada"` (which, granted, does not matter in most of your PRs where you do things early and where we're sure there is no previous value - but as a general principle I prefer assigning nil and be sure there's a reset.)

> that specific case is a fairly gratuitous use of that method as a sort of syntactic sugar

This syntactic sugar assumes you and the reader know `has()` is mostly a check for the thing not being `nil` - and assuming it's usable. I rather have that clear *here* (as it's really simple) than having to look *there* what `has()` really does - because if we are indeed calling it, it must do something more complicated than that :)

Reply:
> > I find that it actually more clearly indicates the intent:
> > e.g., The DocSettings object doesn't contain that key => check elsewhere.
> 
> > Speaking of, (most of) those DocSettings -> G_reader_settings constructs could perhaps be made clearer with something like:
> 
> Yes, that's clearer, in that "DocSettings -> G_reader_settings" case.
> 

Cool will do ;).

> I find it more explicit as it was:
> 
> ```
> 1    self.text_lang_tag = G_reader_settings:readSetting("text_lang_fallback")
> 2    if self.text_lang_tag then
> 3        self.ui.document:setTextMainLang(self.text_lang_tag)
> 4        return
> 5    end
> ```

Fair enough, like I said in the edit, that *was* fairly gratuitous ^^. (And inconsistent, since I left quite a few of those as-is anyway).

> 
> > I'm not sure I get the text_lang comment, because it doesn't affect assignments at all :?
> 
> I meant that _if_ you were to remove/move the step 1 in this sample to not have 2 calls (like you have done elsewhere), you'd get:
> 
> ```
> 2    if G_reader_settings:has("text_lang_fallback") then
> 1        self.text_lang_tag = G_reader_settings:readSetting("text_lang_fallback")
> 3        self.ui.document:setTextMainLang(self.text_lang_tag)
> 4        return
> 5    end
> ```
> 
> You would not be reseting an existing `self.text_lang_tag = "tagada"` (which, granted, does not matter in most of this PR where you do things early and where we're sure there is no previous value - but as a general principle I prefer assigning nil and be sure there's a reset.)

Yeah, I finally figured that's what you meant ^^.
And while I definitely agree with you, if I were to play devil's advocate, I would say that this is Lua: accessing an uninitialized or undefined variable already returns `nil` ;o).

> 
> > that specific case is a fairly gratuitous use of that method as a sort of syntactic sugar
> 
> This syntactic sugar assumes you and the reader know `has()` is mostly a check for the thing not being `nil` - and assuming this means it's usable. I rather have that clear _here_ (as it's really simple) than having to look _there_ what `has()` really does - because if we are indeed calling it, it must do something more complicated than that :)

Well, it just does what it says on the tin ;o).